src/core/artifact.c

```diff
@@ -254,49 +254,74 @@ artifact_checksum_get_or_create(mender_artifact_ctx_t *ctx, const char *filename
-            return MENDER_FAIL;
-        }
+    if (MENDER_OK != mender_sha256_finish(checksum->context, computed)) {
+        mender_log_error("Failed to finish checksum for file '%s'", checksum->filename);
         checksum->context = NULL;
+        return MENDER_FAIL;
+    }
+    checksum->context = NULL;
 
-        if (0 != memcmp(checksum->manifest, computed, MENDER_DIGEST_BUFFER_SIZE)) {
-            mender_log_error("Computed checksum for file '%s' does not match manifest", checksum->filename);
+    if (0 != memcmp(checksum->manifest, computed, MENDER_DIGEST_BUFFER_SIZE)) {
+        mender_log_error("Computed checksum for file '%s' does not match manifest", checksum->filename);
 #if CONFIG_MENDER_LOG_LEVEL >= MENDER_LOG_LEVEL_DBG
-            /* Log the mismatching checksums for debugging */
-            char checksum_str[(MENDER_DIGEST_BUFFER_SIZE * 2) + 1];
+        /* Log the mismatching checksums for debugging */
+        char checksum_str[(MENDER_DIGEST_BUFFER_SIZE * 2) + 1];
 
-            for (int i = 0; i < MENDER_DIGEST_BUFFER_SIZE; i++) {
-                if (2 != snprintf(checksum_str + (i * 2), 3, "%02hhx", checksum->manifest[i])) {
-                    break;
-                }
+        for (int i = 0; i < MENDER_DIGEST_BUFFER_SIZE; i++) {
+            if (2 != snprintf(checksum_str + (i * 2), 3, "%02hhx", checksum->manifest[i])) {
+                break;
             }
-            mender_log_debug("%s: '%s' (manifest)", checksum->filename, checksum_str);
+        }
+        mender_log_debug("%s: '%s' (manifest)", checksum->filename, checksum_str);
 
-            for (int i = 0; i < MENDER_DIGEST_BUFFER_SIZE; i++) {
-                if (2 != snprintf(checksum_str + (i * 2), 3, "%02hhx", computed[i])) {
-                    break;
-                }
+        for (int i = 0; i < MENDER_DIGEST_BUFFER_SIZE; i++) {
+            if (2 != snprintf(checksum_str + (i * 2), 3, "%02hhx", computed[i])) {
+                break;
             }
-            mender_log_debug("%s: '%s' (computed)", checksum->filename, checksum_str);
-#endif /* CONFIG_MENDER_LOG_LEVEL >= MENDER_LOG_LEVEL_DBG */
-            return MENDER_FAIL;
         }
+        mender_log_debug("%s: '%s' (computed)", checksum->filename, checksum_str);
+#endif /* CONFIG_MENDER_LOG_LEVEL >= MENDER_LOG_LEVEL_DBG */
+        return MENDER_FAIL;
     }
 #endif /* CONFIG_MENDER_FULL_PARSE_ARTIFACT */
     return MENDER_OK;
 }
 
+mender_err_t
+mender_artifact_check_integrity_item(mender_artifact_ctx_t *ctx, const char *filename) {
+    assert(NULL != ctx);
+    bool found = false;
+    for (mender_artifact_checksum_t *checksum = ctx->artifact_info.checksums; NULL != checksum && !found; checksum = checksum->next) {
+        if (StringEqual(filename, checksum->filename)) {
+            found             = true;
+            checksum->checked = true;
+            if (MENDER_OK != is_checksum_valid(checksum)) {
+                return MENDER_FAIL;
+            }
```

Comment:
Why not `return is_checksum_valid(checksum);` instead of the whole `if`? With that you could also do `return MENDER_FAIL` below because it would only be reached when not found. There cannot be multiple checksums for the same file name, right?